doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -90,11 +93,63 @@ Rewriting with Leibniz and setoid equality
      any of them can be rewritten.  If not specified, only the first occurrence
      in the conclusion is replaced.
 
-     If :n:`at @occs_nums` is specified, rewriting is always done with
-     :ref:`setoid rewriting <generalizedrewriting>`, even for Leibniz’s equality,
-     which means that you must `Require Setoid` to use that form.
-     However, note that :tacn:`rewrite` (even when using setoid rewriting) and
-     :tacn:`setoid_rewrite` don't behave identically (as already mentioned above).
+     .. note::
+
+        If :n:`at @occs_nums` is specified, rewriting is always done
+        with :ref:`setoid rewriting <generalizedrewriting>`, even for
+        Leibniz equality, which means that you must `Require
+        Setoid` to use that form.  However, note that :tacn:`rewrite`
+        (even when using setoid rewriting) and :tacn:`setoid_rewrite`
+        don't behave identically (as already mentioned above).
+
+     .. note::
+
+        For each selected hypothesis (and the conclusion),
+        :tacn:`rewrite` will instantiate the rewriting lemma first
+        before taking the `at` clause into account, which limits the
+        number of occurrences that can be rewritten.  You can force
+        :tacn:`rewrite` to instantiate the rewriting lemma differently
+        by using a `with` clause (see :n:`@one_term_with_bindings`) or
+        you can use :tacn:`setoid_rewrite`, which considers each
+        occurrence separately.
+
+        .. example::
+
+           .. coqtop:: none
+
+              Require Import Arith.
+
+           If we want to rewrite the right-hand side, it will not work
+           like this:
```

Comment:
```suggestion
           While rewriting the left-hand side works with `at 1`, if we want to rewrite the right-hand side,
           this doesn't work:
```

Seems like a bug that we should fix soon.

Reply:
That's not exactly a bug. See the discussions linked from this PR.

Reply:
I read the issue.  Why do you not think it's a bug?  It's unexpected behavior. 
Shouldn't we try to state when/why it fails rather than just say it fails for this single example?  Of course, your changes are a step forward.

Reply:
I read the issue.  Why do you not think it's s bug?  Would be better to explain how it actually works rather than just a single example.  Occurrence clauses should behave consistently.  It should be possible to do a deprecation and users can get the old behavior with explicit bindings. IMHO.  What you've added is a step forward b of course.

Reply:
Well, I did think this was a bug initially. This is why I opened the issue. It's no longer a bug if developers tell me this is according to the spec, even if the spec is weird. I documented the spec. Now, the question becomes whether we want to change the spec to have more consistent behavior. I would tend to think so but others seem to think it would be too impactful. The only way to know is to test. On the other hand, I'm not keen to start doing that now given that it's not clear to me what the future of `rewrite` vs `setoid_rewrite` vs SSReflect's `rewrite` (which is often claimed to be better) is.

>Would be better to explain how it actually works rather than just a single example.

I did explain quite precisely what happens. I only added an example because Guillaume asked for one. But the explanation is still there, above the example:

>For each selected hypothesis (and the conclusion), `rewrite` instantiates the rewriting lemma first before taking the `at` clause into account, which limits the number of occurrences that can be rewritten.

Reply:
> It's no longer a bug if developers tell me this is according to the spec, even if the spec is weird.

Well, no.  If the spec is weird for no good reason, that is a bug.  Bugs can be a discrepancy at the level of spec versus behavior (objective) OR spec versus usability (subjective).  If the spec says ```The :tacn:`bombs away` tactic formats your hard drive```, that's absurd and absolutely a bug.

If you're willing, it would be great to evaluate the impact of the change and see how much stuff breaks and how hard it is to fix--after inquiring about the future of the various rewrite tactics you mention.  If we want to make it easy for users, we could do the matching both ways and give a warning when they differ.  Or create a `rewrite old` variant with the current behavior.

Reply:
>If you're willing, it would be great to evaluate the impact of the change and see how much stuff breaks and how hard it is to fix--after inquiring about the future of the various rewrite tactics you mention.

It's going to be a long and complex process to understand, decide and act to have these three tactics become one. Until this is the case, it's not clear how much effort should be spent on changing their current behavior. So no, I'm not really willing to invest my own time in this now.

Reply:
That's fine, not surprising.